openpos-flow/src/main/java/org/jumpmind/pos/core/content/AbstractFileContentProvider.java

```diff
@@ -98,26 +100,52 @@ public String getMostSpecificContent(String deviceId, String key, String baseCon
         return Collections.emptyList();
     }
 
-    protected List<String> getPossibleContentDirs(String key) {
-        List<String> possibleContentDirs = new ArrayList<>();
+    protected List<String> getAllPossibleContentDirPermutations(String key) {
+        List<String> permutations = new ArrayList<>();
+
         if (providerProperties != null && personalizationProperties != null) {
-            for (int prop = providerProperties.length; prop >= 0; prop--) {
-                StringBuilder builder = new StringBuilder(key);
+            getPermutations(permutations, providerProperties, 0);
 
-                for (int i = 0; i < prop; i++) {
-                    String property = providerProperties[i];
-                    if (personalizationProperties.containsKey(property)) {
-                        builder.append("/").append(personalizationProperties.get(property));
+            List<String> additions = new ArrayList<>();
+            permutations.forEach(s -> {
+                for(int x = 1; x < s.length(); x++) {
+                    if(s.charAt(x) == '/' && !additions.contains(s.substring(0, x) + "/")) {
+                        additions.add(s.substring(0,x) + "/");
                     }
                 }
+            });
 
-                possibleContentDirs.add(builder.toString());
+            additions.sort((s1, s2) -> s2.split("/").length - s1.split("/").length);
+            permutations.addAll(additions);
+            permutations = permutations.stream().map(s -> key + s).collect(Collectors.toList());
+        }
+        return permutations;
+    }
+
+    private void getPermutations(List<String> values, String[] array, int pos){
```

Comment:
Could we use better variable names? I had to scroll back and forth to try and straighten out what array was and pos makes me think of point-of-sale
```suggestion
    private void getPermutations(List<String> values, String[] provderProperties, int position){
```